### ISSUE TITLE
DAL > MSSQL2Adapter > ILIKE invalid syntax

### ISSUE THREAD

**Issue by niphlod** [web2py/web2py#785]
_From [score2..._at_gmx.net](https://code.google.com/u/100357035028011448959/) on February 05, 2015 11:36:30_

1. MSSQL2Adapter.represent(value) returns N'value' for input type 'string', 'text', 'json'

2. MSSQL2Adapter.ILIKE(first, second) expands the second argument and then lowers it, leading to n'value' which is invalid syntax for mssql

Possible fix:
    def ILIKE(self, first, second):
        """Case insensitive like operator"""
        return '(LOWER(&#37;s) LIKE &#37;s)' &#37; (self.expand(first),
                                 self.expand(second.lower(), 'string')) What version of the product are you using? On what operating system? 2.9.12

_Original issue: http://code.google.com/p/web2py/issues/detail?id=2044_

**Comment by ilvalle** [comment]
I'm not convinced by the proposed fix. the parameter second can be anything not only a string.
Could you post an example of your query ?

**Comment by niphlod** [comment]
the problem is subtle ... mssql2 is a shallow try of supporting unicode fields in mssql, and the statement for string and text fields to work needs a 'N' prependend to the comparison value (if the comparison value is a string or a text)
i.e.
where field = N'something'

that "N" should be always uppercase.

**Comment by skoa-h** [comment]
Here is a simple example of the MSSQL2Adapter breaking for like() and contains() using non-case-sensitivity.


```
from gluon.dal import DAL, Field

db = DAL('mssql2://USER:PW@SERVER/DB')

db.define_table('bu', Field('title', 'string'))

db.define_table('activity', Field('business_unit', 'list:reference bu'))

# working
query = db.bu.title.like("foo", case_sensitive=True)
print query # (bu.title LIKE N'foo')
rows = db(query).select()

query = db.activity.business_unit.contains(1, case_sensitive=True)
print query # (activity.business_unit LIKE N'%|1|%')
rows = db(query).select()

# Failing
try:
    query = db.bu.title.like("foo", case_sensitive=False)
    print query # (LOWER(bu.title) LIKE n'foo')
    rows = db(query).select()
except Exception, e:
    print e

try:
    query = db.activity.business_unit.contains(1, case_sensitive=False)
    print query # (LOWER(activity.business_unit) LIKE n'%|1|%')
    rows = db(query).select()
except Exception, e:
    print e

**Comment by niphlod** [comment]
ok, kinda got it. Thanks to careful inspection, it works for ANY case except a single one .... but it's pydal's logic that is convoluted to a point that for the mssql adapter is really hard to reconcile.
pydal's order of operations seems to be the inverse of what "should happen"...

Directly from the unittest that fails....
```python
db = DAL('sqlite://storage.sqlite')
db.define_table('tt', Field('aa', 'list:string'), Field('bb','string'))
print db(db.tt.aa.contains(db.tt.bb))._count()

>>> SELECT count(*) FROM tt WHERE (LOWER(tt.aa) LIKE 
(('%|' || (replace((replace(tt.bb,'%','%%')),'|','||')) || '|%')));
```

Note the "replace" not being "REPLACE".... The thing here is, for a contains() that needs to be NOT case-sensitive, the "trick" in pydal is calling LOWER() on the first, and then lower()ize (in python) the second part. This trick works perfectly UNTIL you lower()ize something that is an expression. In most dialects this works, but for MSSQL, whose "unicode-aware" query that can work should be
```
SELECT count(*) FROM tt WHERE (LOWER(tt.aa) LIKE 
((N'%|' + (REPLACE((REPLACE(tt.bb,N'%',N'%%')),N'|',N'||')) + N'|%')));
```
instead of the (even if patched)
```
SELECT count(*) FROM tt WHERE (LOWER(tt.aa) LIKE ((n'%|' + (replace((replace(tt.
bb,n'%',n'%%')),n'|',n'||')) + n'|%')));
```
that is now.

tl;dr : calling self.expand() on an Expression and then lower()izing it in python (even if the resulting command is parseable for most backends) isn't really "fail-proof" (or so I think). mssql just fails because ``n'`` is not parseable as ``N'`` and I can work around in ILIKE not lower()izing if ``second`` is an Expression, but I feat that other backends needs a revisitation of this method too.

tl;dr, v2: is ILIKE in base.py inherently broken because it lower()ize even an Expression or not ?

@ilvalle , @gi0baro : opinions ?

**Comment by ilvalle** [comment]
We should revises how lower()/upper() in python are applied to Expressions. 
We should demand those operations to their arguments (first/second), if an argument is an Expression recursively demand to its arguments and so on. I'll have a look to see how it is feasible

**Comment by gi0baro** [comment]
@niphlod @ilvalle agree to review lower()/upper() on `Expression` arguments

**Comment by ilvalle** [comment]
it's not so easy. The starting point should be replacing
```
def ILIKE(self, first, second):
     """Case insensitive like operator"""
     return '(LOWER(%s) LIKE %s)' % (self.expand(first), self.expand(second, 'string').lower())
```
with
```
def ILIKE(self, first, second):
     """Case insensitive like operator"""
     return '(LOWER(%s) LIKE %s)' % (self.expand(first), self.expand(second.lower(), 'string'))
```
because lowering an expanded query can leads to some errors. 
However, if second is a string, everything is fine, but if second is an Expression, ```Expression.lower()``` will be called which isn't lower() at python level but rather lower at sql level. This later leads to the error highlighted by @skoa-h and @niphlod. 
In my view, ```Expression.lower()``` should lower()ize at python level its arguments.
The original ```Expression.lower()``` which defines the sql LOWER should be renamed. 
Could this work? Don't know the impact that this approach can have.

**Comment by niphlod** [comment]
this has apparently been fixed in the latest pydal package.
